pymc/gp/hsgp.py

```diff
@@ -0,0 +1,376 @@
+#   Copyright 2023 The PyMC Developers
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+import warnings
+
+from types import ModuleType
+from typing import Optional, Sequence, Tuple, Union
+
+import numpy as np
+import pytensor.tensor as pt
+
+import pymc as pm
+
+from pymc.gp.cov import Covariance
+from pymc.gp.gp import Base
+from pymc.gp.mean import Mean, Zero
+
+TensorVariable = Union[np.ndarray, pt.TensorVariable]
+TensorConstant = Union[np.ndarray, pt.TensorConstant]
+
+
+def set_boundaries(
+    Xs: TensorVariable,
+    L: Optional[Sequence] = None,
+    c: Optional[Union[float, Sequence]] = None,
+    tl: ModuleType = np,
+) -> Tuple[TensorVariable, TensorVariable, Union[TensorConstant, float]]:
+    """R Compute the boundary over which the approximation is accurate using the centered input data
+    locations, `Xs`.  It is requried that `Xs` has a mean at zero for each dimension, such that
+    `np.mean(Xs, axis=0) == [0, ..., 0]`.  If `L` is provided, c is calculated instead.  `tl` stands
+    for "tensor library", so the user can choose whether basic calculations are done with pytensor
+    or numpy.
+
+    Parameters
+    ----------
+    Xs: array-like
+        Function input values.  Assumes they have been mean subtracted or centered at zero.
+    """
+
+    if tl.__name__ not in ("numpy", "pytensor.tensor"):
+        raise ValueError("tl must be either numpy or pytensor.tensor.")
+
+    S = tl.max(tl.abs(Xs), axis=0)
+
+    if L is None and c is not None:
+        L = c * S
+    elif c is None and L is not None:
+        c = L / S
+    elif c is None and L is None:
+        raise ValueError("At least one of `c` or `L` must be supplied.")
+    # if both are passed, L takes precedent.
+
+    if tl.__name__ == "numpy":
+        L = tl.asarray(L)
+    elif tl.__name__ == "pytensor.tensor":
+        L = tl.as_tensor_variable(L)
+
+    return S, L, c
+
+
+def calc_eigenvalues(L: TensorConstant, m: Sequence[int], tl: ModuleType = np):
+    """R Calculate eigenvalues of the Laplacian."""
+    S = np.meshgrid(*[np.arange(1, 1 + m[d]) for d in range(len(m))])
+    S = np.vstack([s.flatten() for s in S]).T
+    return tl.square((np.pi * S) / (2 * L))
```

Comment:
```suggestion
    S = np.meshgrid(*[np.arange(1, 1 + m[d]) for d in range(len(m))])
    Sarr = np.vstack([s.flatten() for s in S]).T
    return tl.square((np.pi * Sarr) / (2 * L))
```
The first assignment defines the type (here: `list`) and that's why the second assignment gives an error.
```
Incompatible types in assignment (expression has type "ndarray[Any, dtype[Any]]", variable has type "List[ndarray[Any, dtype[Any]]]")
Unsupported operand types for * ("float" and "List[ndarray[Any, dtype[Any]]]")
```
By creating a new variable the type conflicts should be resolved.